sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryParallelReadTableSource.java

```diff
@@ -0,0 +1,210 @@
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      Coder<T> coder,
+      BigQueryServices bqServices,
+      ReadSessionOptions readSessionOptions) {
+    return new BigQueryParallelReadTableSource<>(
+        NestedValueProvider.of(
+            checkNotNull(tableRefProvider, "tableRefProvider"),
+            new TableRefToJson()),
+        parseFn,
+        coder,
+        bqServices,
+        readSessionOptions,
+        null,
+        null,
+        null);
+  }
+
+  private BigQueryParallelReadTableSource(
+      ValueProvider<String> jsonTableRefProvider,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      Coder<T> coder,
+      BigQueryServices bqServices,
+      ReadSessionOptions readSessionOptions,
+      Session readSession,
+      ReadLocation readLocation,
+      Long readSizeBytes) {
+    this.jsonTableRefProvider = checkNotNull(jsonTableRefProvider, "jsonTableRefProvider");
+    this.parseFn = checkNotNull(parseFn, "parseFn");
+    this.coder = checkNotNull(coder, "coder");
+    this.bqServices = checkNotNull(bqServices, "bqServices");
+    this.readSessionOptions = readSessionOptions;
+    this.readSession = readSession;
+    this.readLocation = readLocation;
+    this.readSizeBytes = readSizeBytes;
+  }
+
+  /**
+   * Gets the estimated number of bytes returned by the current source. For sources created by a
+   * split() operation, this is a fraction of the total table size.
+   */
+  @Override
+  public synchronized long getEstimatedSizeBytes(PipelineOptions options)
+      throws IOException, InterruptedException {
+    if (readSizeBytes != null) {
+      return readSizeBytes;
+    }
+
+    if (cachedReadSizeBytes == null) {
+      TableReference tableReference = BigQueryIO.JSON_FACTORY.fromString(
+          jsonTableRefProvider.get(), TableReference.class);
+      Table table = bqServices.getDatasetService(options.as(BigQueryOptions.class))
+          .getTable(tableReference);
+      cachedReadSizeBytes = table.getNumBytes();
+    }
+
+    return cachedReadSizeBytes;
+  }
+
+  @Override
+  public List<BoundedSource<T>> split(long desiredBundleSizeBytes, PipelineOptions options)
+      throws IOException, InterruptedException {
+    if (readSession != null) {
+      return ImmutableList.of(this);
+    }
+
+    CreateSessionRequest.Builder requestBuilder = CreateSessionRequest.newBuilder();
+    TableReference tableReference = BigQueryIO.JSON_FACTORY.fromString(
+        jsonTableRefProvider.get(), TableReference.class);
+    requestBuilder.setTableReference(
+        TableReferenceProto.TableReference.newBuilder()
+            .setProjectId(tableReference.getProjectId())
+            .setDatasetId(tableReference.getDatasetId())
+            .setTableId(tableReference.getTableId()));
+
+    Long tableSizeBytes = getEstimatedSizeBytes(options);
+    requestBuilder.setReaderCount(
+        (tableSizeBytes / desiredBundleSizeBytes) > Integer.MAX_VALUE
```

Comment:
Please add some safeguards here.
desiredBundleSizeBytes may be 0 in some cases (at least I can't guarantee that it won't be).
Also it may be very small or very large. One case where I've seen it be very small: BigQuery mis-estimates size of a huge table and Dataflow gives a small bundle size, then during splitting we end up with an enormous number of bundles. There should be an upper limit, e.g. 10k splits or so. A lower limit can't hurt too (if size estimation fails, Dataflow will give some default bundle size that may be too large).

(yes, this is super funky - I think we'd be better off if size estimation and splitting hints didn't exist at all, and we're moving towards a world in which they don't)

Reply:
I will add the safeguards you describe. What's a reasonable lower bound?

Reply:
10 should be reasonable.

Reply:
Done.